src/fsharp/opt.fs

```diff
@@ -1300,7 +1298,7 @@ and OpHasEffect g op =
     | TOp.TupleFieldGet(_) -> false
     | TOp.ExnFieldGet(ecref,n) -> isExnFieldMutable ecref n 
     | TOp.RefAddrGet -> false
-    | TOp.ValFieldGet rfref  -> rfref.RecdField.IsMutable
+    | TOp.ValFieldGet rfref  -> rfref.RecdField.IsMutable || HasFSharpAttribute g g.attrib_AllowNullLiteralAttribute rfref.Tycon.Attribs
```

Comment:
I think this is more accurate (AllowNullLiteralAttribute can be used with a "false" argument, at least on .NET types, it's better to be explicit that we're searching for a default or explicit "true"):

    TryFindTyconRefBoolAttribute g m g.attrib_AllowNullLiteralAttribute rfref.TyconRef = Some(true)